File: 0003_Longest_Substring_Without_Repeating_Characters.py
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        if len(s) == 0: return 0
        start, end = 0, 1
        res = 1
        for i in range(1, len(s)):
            while s[i] in s[start:end]:
                start += 1
            end += 1
            res = max(res, end-start)
        return res

# 2019/03/06 update:
class Solution:
    def lengthOfLongestSubstring(self, s):
        """
        :type s: str
        :rtype: int
        """
        i, j = 0, 0
        exist = {}
        result = 0
        while j < len(s):
            if s[j] not in exist:
                exist[s[j]] = 1
            else:
                while i < j and s[i] != s[j]:
                    exist.pop(s[i])
                    i += 1
                i += 1
            j += 1
            result = max(result, j-i)
            
        return result

File: 0003_Longest_Substring_Without_Repeating_Characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        int result = 0;
        int i = 0, j = 0;
        HashSet<Character> set = new HashSet<Character>();
        while (j < s.length()) {
            while (set.contains(s.charAt(j))) {
                set.remove(s.charAt(i++));
            }
            set.add(s.charAt(j++));
            result = Math.max(result, set.size());
        }
        return result;
    }
}